shuffle/sequences.go

```diff
@@ -0,0 +1,155 @@
+package shuffle
+
+import (
+	"crypto/cipher"
+	"fmt"
+
+	"go.dedis.ch/kyber/v3"
+	"go.dedis.ch/kyber/v3/proof"
+)
+
+// SequencesShuffle shuffles a sequence of ElGamal pairs based on Section 5 of
+// "Verifiable Mixing (Shuffling) of ElGamal Pairs" by Andrew Neff (April 2004)
+//
+// The function expects X and Y to be the same dimension, with each row having
+// the same length. It also expect X and Y to have at least one element.
+//
+// Dim X and Y: [<sequence length, j>, <number of sequences, i>]
+//
+// The number of rows defines the sequences length. The number of columns
+// defines the number of sequences.
+//
+//  Seq 1  Seq 2  Seq 3
+//  (0,0)  (0,1)  (0,2)
+//  (1,0)  (1,1)  (1,2)
+//  (2,0)  (2,1)  (2,2)
+//
+// In the code coordinates are (j,i), where 1 ≤ i ≤ k, 1 ≤ j ≤ NQ
+//
+// Last coordinate is (NQ-1, k-1)
+//
+// Variable names are as representative to the paper as possible. Instead of
+// representing (variable name with a bar on top), such as (X with a bar on top)
+// with Xbar, we represent it with a repeating letter, such as XX
+func SequencesShuffle(group kyber.Group, g, h kyber.Point, X, Y [][]kyber.Point,
+	rand cipher.Stream) (XX, YY [][]kyber.Point, getProver func(e []kyber.Scalar) (
+	proof.Prover, error)) {
+
+	NQ := len(X)
+	k := len(X[0])
+
+	// Pick a random permutation used in ALL k ElGamal sequences. The permutation
+	// (π) of an ElGamal pair at index i always outputs to the same index
+	pi := make([]int, k)
+	for i := 0; i < k; i++ {
+		pi[i] = i
+	}
+
+	// Fisher–Yates shuffle
+	for i := k - 1; i > 0; i-- {
+		j := int(randUint64(rand) % uint64(i+1))
+		if j != i {
+			pi[i], pi[j] = pi[j], pi[i]
+		}
+	}
+
+	// Pick a fresh ElGamal blinding factor β(j, i) for each ElGamal sequence
+	// and each ElGamal pair
+	beta := make([][]kyber.Scalar, NQ)
+	for j := 0; j < NQ; j++ {
```

Comment:
Confusing to me: I usually start with `i` in the outer loop, and then go through `j`, `k`, ... in the inner loops. Except if the paper uses the same notation.

Here and in other places.

Reply:
Yes, it follows the paper's notation. But still we could iterate over the rows first, I agree it's quite confusing, and probably not optimized.

Reply:
Yes, it follows the paper's notation and I agree it's confusing. We choose to stick to the paper's notation.

Reply:
OK - didn't look at the paper. You could even add an URL to arxive, if it's in there.